db/system_keyspace.cc

```diff
@@ -99,6 +99,10 @@ namespace {
             props.enable_schema_commitlog();
         }
     });
+
+    tombstone_gc_options get_group0_tombstone_gc_options() {
+        return tombstone_gc_options({{"mode", "repair"}, {"propagation_delay_in_seconds", "10"}});
```

Comment:
Group zero managed tables aren't repaired, so how will this work?

Reply:
Didn't we discuss it already. The mode name is "repair", but the only thing repair does is set a timestamp for the oldest tombstone to be preserved. With group this timestamp setting moves from repair to another module, but the mode name is already an ABI.

Reply:
Create a new name. The new name should not include the word "repair". You're confusing implementation and interface here. The implementation of the two modes can be the same (although, it won't be the same, since the source of the timestamp is different). The names have to be different.

Reply:
I am not confusing anything. The patch does not introduce new tombstone gc mode. It uses existing implementation and interface. It so happens that the naming of the existing interface is not of your liking because if is derived from the only user of the interface until now. Do you propose to introduce an alias? I do not see how it improves anything, but OK.

Reply:
No. tombstone_gc = repair tells the user that tombstones will be garbage-collected after repair.

If we rename it to tombstone_gc = after_some_timestamp_determined_externally, it works equally well for raft group 0 and regular user tables, but doesn't tell the user anything about how tombstone garbage collection is managed.

If we keep it as tombstone_gc = repair, then it's accurate about user tables, but tells the user that group 0 tables will garbage collect tombstones after repair, which is incorrect.

Reply:
> No. tombstone_gc = repair tells the user that tombstones will be garbage-collected after repair.

What is "repair"? Is there guaranty that there is only one kind of a repair? What if we have different algorithms for repair for small and large tables? Should a user be aware of the difference and have different configuration for the mode? What will happen when repairs will be automated and hidden from a user what "mode repair" will tell the user then? Why do you assume that raft's catch up mechanism is not just a way to automatically repair raft managed tables? After all isn't "repair" is a process  that guaranties that all nodes saw all data written before a certain time point?

> 
> If we rename it to tombstone_gc = after_some_timestamp_determined_externally, it works equally well for raft group 0 and regular user tables, but doesn't tell the user anything about how tombstone garbage collection is managed.
> 
> If we keep it as tombstone_gc = repair, then it's accurate about user tables, but tells the user that group 0 tables will garbage collect tombstones after repair, which is incorrect.

Why is it incorrect? What is a "repair" for you? A bunch of code that we have right now in the repository under "repair" directory? Any change to the directory makes it a different kind of the repair?

Like Korolev once said (paraphrased):  "If you criticize then propose an alternative if you propose an alternative - work on it". You seams to be stuck on criticize. What do you propose?

Reply:
The option already discussed before is to introduce an alias, but the question is if that really makes a practical sense, as:
* the alias would (at least currently) do exactly the same what the current code using the GC mode "repair" does
* it will complicate the code
* and this all is not visible to (or controlled by) the user (the group0 tables are internal, the users can eventually desc them if they know about their existence, but I'd assume that is not a regular usage - perhaps only in some recovery scenarios or alike)

Is there any other possibility beside an alias @avikivity you have in mind? And if the alias is the right proposition, do you insist on it even considering all the above (and the Gleb's comment)?

We'd also need to think about what a good name for such alias would be ... the current values are:  "timeout", "disabled", "immediate", "repair". Could be perhaps something like GC mode = "consistent" / "sync" / "reconcile" ?

Reply:
How about simply putting a big ugly comment inside `get_group0_tombstone_gc_options` above this line that explains why `repair` mode is the thing that we're using and why the name is not accurate?

Reply:
My proposal: tombstone_gc = 'group0-managed'

Reply:
> repair is the procedure exercised by `nodetool repair`. There's no reason to go metaphysical on it.
> 

What about automatic repair that we are working on?

Reply:
> My proposal: tombstone_gc = 'group0-managed'

And when we will have regular tables managed by raft we will introduce one more name for the same thing?

Reply:
> Users will not read the comment inside get_group0_tombstone_gc_options.

@avikivity Users don't care, this is for internal system tables.

Reply:
> the alias would (at least currently) do exactly the same what the current code using the GC mode "repair" does

It's not nearly exactly. One of them uses the last repair time to derive the gc timestamp. The other uses some gossip disseminated timestamp.

> it will complicate the code

Please estimate the complexity. How many weeks will it take to assign a new name?

> and this all is not visible to (or controlled by) the user (the group0 tables are internal, the users can eventually desc them if they know about their existence, but I'd assume that is not a regular usage - perhaps only in some recovery scenarios or alike)

Users describe and use system tables all the time. They are a public interface and are not internal.

Reply:
> What about automatic repair that we are working on?

And what about the repair that is executed during bootstrap with RBNO

Reply:
> > My proposal: tombstone_gc = 'group0-managed'
> 
> And when we will have regular tables managed by raft we will introduce one more name for the same thing?

That's what I was wondering too, if we introduce the alias whether we could give it a name that might actually apply to all similar cases (maybe even to the "repair" case). As all of these basically means tombstone GC up to the point of where all the tables (or replicas) are in sync (consistent), that's why I was looking into the synonyms of sync / catched up etc. (thus mode = "consistent" / "reconcile" or alike)

This could then apply to all similar synchronization scenarios (whether it is done by the group0 / raft catching up, or synchronized by any different means, e.g the actual nodetool repair) - then in fact the "repair" would be even redundant and could in fact become the alias of the new mode (but I'd not remove it as it is already established).